src/hash_generator/dlNormalizer.js

```diff
@@ -0,0 +1,398 @@
+/**
+ * (c) Copyright Reserved EVRYTHNG Limited 2021. All rights reserved.
+ * Use of this material is subject to license.
+ * Copying and unauthorised use of this material strictly prohibited.
+ */
+
+/* eslint-disable no-useless-escape */
+
+/**
+ * it protects some character with '%..'
+ * @param {string} input - the string to protect
+ * @return {string} the updated string
+ */
+export const webUriPercentEncoder = (input) => input.split('!').join('%21')
+  .split('(').join('%28')
+  .split(')')
+  .join('%29')
+  .split('*')
+  .join('%2A')
+  .split('+')
+  .join('%2B')
+  .split(',')
+  .join('%2C')
+  .split(':')
+  .join('%3A');
+
+/**
+ * @param {string} input - id that requires checkdigit
+ * @param {boolean} throwError - if set to true, it will throw an error if the event misses some
+ * fields for example. Otherwise, it won't throw an error and it will still return the generated id
+ * @return {number} the checkDigit of the parameter
+ */
+const getCheckDigit = (input, throwError) => {
+  try {
+    const array = input.split('').reverse();
+    let total = 0;
+    let i = 1;
+    array.forEach((number) => {
+      const n = parseInt(number, 10);
+      if (i % 2 === 0) {
+        total += n;
+      } else {
+        total += n * 3;
+      }
+      i += 1;
+    });
+    return Math.ceil(total / 10) * 10 - total;
+  } catch (e) {
+    if (throwError) {
+      throw new Error(`There was a problem during the check digit operation of ${input}`);
+    }
+    return 0;
```

Comment:
While we may ignore a problem generating a check digit, other systems that validate GTINs with check digits may not. Suggest we _always_ error here if the input is bad.

Reply:
I see your point. But the goal of the `throwError` boolean is to choose whether the generation of an ID can throw an error or not. If the developer wants to have a valid ID that follows the algorithm and the EPCIS rules, it will set `throwError` to true and this function would throw. But if a developer only wants to generate an ID and don't care if there are any problems in its gtin, he should be able to do it.

Of course, by default, `throwError` is set to true.